PROTOCOL.md

```diff
@@ -0,0 +1,193 @@
+  runtime environment. While the server should not use this
+  information for request processing, this header is heavily used
+  for analytics and troubleshooting purposes.
+
+* **RPC-Timeout** header indicates the client-specified request
+  timeout in seconds, such as "10". If **RPC-Timeout** is omitted, the
+  server should use a pre-configured timeout value, by default it
+  should be 10 seconds.
+
+The **Request-Body** is the encoded request message, contained in the
+HTTP request body. The encoding is specified by the `Content-Type`
+header.
+
+### Responses
+
+**Response ::= Response-Headers Response-Body**
+
+The **Response-Headers** are just normal HTTP response headers. The
+Twirp wire protocol uses the following headers.
+
+* **Content-Type** The value should be either "application/x-protobuf"
+  or "application/json" to indicate the encoding of the response
+  message. It must match the "Content-Type" header in the request.
+
+The **Request-Body** is the encoded response message contained in the
+HTTP response body. The encoding is specified by the `Content-Type`
+header.
+
+## Example
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com".
+
+```proto
+package twirp;
+
+service Echo {
+  rpc Hello(HelloRequest) returns (HelloResponse);
+}
+
+message HelloRequest {
+  string message;
+}
+
+message HelloResponse {
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/x-protobuf
+Content-Length: 15
+
+<encoded HelloRequest>
+```
+
+**JSON Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+**Proto Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/x-protobuf
+Content-Length: 15
+
+<encoded HelloResponse>
+```
+
+**JSON Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+## Errors
+
+If an error occurs when the server processes a request, the server
+must return an error payload as the response message, and correctly
+set the HTTP status code. Please see
+[`google.rpc.Code`](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)
+on how to map typical server errors to HTTP status codes.
+
+### Timeout errors
+
+For a single request, there is a client-specified timeout and a
+server-configured timeout. If a request misses the server-configured
+timeout, the server must return a `503` error. If a request misses
+the client-specified timeout that it is shorter then the server-
+configured timeout, the server must return a `504` error.
+This allows more accurate measurement of the server availability.
+
+### Network errors
+
+If a client fails to reach the server due to network errors, the
```

Comment:
I don't think we should specify this. I don't think client libraries should be exposing HTTP status codes, just Twirp error codes.